---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-"
)
```
# ANNOmtDNA
### Association analysis of heteroplasmic mtDNA mutations.

**Authors:** Xianbang Sun (maintainer, sxb3000@bu.edu), Katia Bulekova, Chunyu Liu, Jessica L. Fetterman, Jun Ding, Kaiyu Yan, Daniel Levy<br>
**Date:** "`r format(Sys.Date(), '%m/%d/%Y')`"

<!-- badges: start -->
<!-- badges: end -->


  ANNOmtDNA is an R package for identifying and annotating mtDNA sequence variations. This package allows to identify mtDNA sequence variations with user-specified thresholds, and provide several plots to visualize mtDNA sequence variations. It also annotates all mtDNA variations and predicts functionality for mtDNA variants in the coding and tRNA regions. 
  
*Note:* if you use ANNOmtDNA in published research, please cite:

Xianbang Sun, Katia Bulekova, Chunyu Liu, Jessica L. Fetterman, Jun Ding, Kaiyu Yan, Daniel Levy (2020)
mtdnaANNO: an R package for comprehensive annotation of mtDNA sequence variation


## Installation

You can install the  development version of ANNOmtDNA package from [GitHub](https://github.com/) with:

``` {r install}
# install.packages("devtools")
# devtools::install_github("mtDNA-BU/ANNOmtDNA")
library(ANNOmtDNA)
```
## Standard workflow

Import the original allele, frequency and coverage datasets
```{r path_def_user, include = TRUE, eval=FALSE}
input_path = "/path/to/input/directory/"
```
```{r path_def, include = FALSE}
input_path = "/restricted/projectnb/mtdna-alcohol/Sun_Xianbang/ARIC/"
```



```{r chunk 2}
allele <- read.csv(paste0(input_path, "allele/allele_ANNOmtDNA.csv"), header = T, stringsAsFactors=FALSE,
                   colClasses = c("character"))
freq <- read.csv(paste0(input_path,"freq/freq_ANNOmtDNA.csv"), header=T, stringsAsFactors=FALSE,
                      colClasses = c("character"))
coverage <- read.csv(paste0(input_path,"coverage/coverage_ANNOmtDNA.csv"), header=T)
allele <- as.matrix(allele)
freq <- as.matrix(freq)
coverage <- data.matrix(coverage)
```


#### Compute alternative allele fraction (AAF) by mtAAF function
```{r chunk 5}
AAF <- mtAAF(allele, freq)
```

Scatter plot of output of mtAAF function, each point is AAF of a subject at a locus, the x axis is the mtDNA loci and the y axis is the range of AAF (0-100%)
```{r scatterPlot, fig.height=4, fig.width=8}
plot(AAF)
```

Plot the mean coverage of each locus by plotCover function, the x axis is the mtDNA loci and y axis is the mean coverage
```{r plotMeanCoverage, fig.height=4, fig.width=8}
plotCover(coverage)
```

Histogram of the mean coverage of each subject by histSampCov function
```{r histMeanCoverage, fig.height=5, fig.width=5}
histSampCov(coverage)
```

#### Summarize and annotate mtDNA mutations by mtSummary function
Specify the path to output the annotation file and histograms of mtDNA mutation burden of subjects and number of mutations carried by each loci for heteroplasmic and homoplasmic mutations

Users can specify lower bound and upper bound of the threshold by thre.lower and thre.upper arguments. By default, thre.lower=0.03 and thre.upper=0.97. That is, if $0.03\leq AAF\leq 0.97$, it is a heteroplasmic mutation; and if $AAF>0.97$, it is a homoplasmic mutation. Users can also choose different gene regions by loci argument. For example, to choose tRNA region, set loci="tRNA". Users can also choose types of mutations to be annotated by type argument. For example, to choose heteroplasmic mutations, set type="heter", and choose homoplasmic mutations, set type="homo".
Run the mtSummary function, only include loci of coding region, annotate for both of heteroplasmic and homoplasmic mutations.

```{r chunk 9, echo=TRUE, eval=FALSE}
 output_path <- "/output/dir/"
```
```{r chunk 10, echo=F}
output_path <- paste0(tempdir(),"/")
```

```{r chunk 11}
mtSum <- mtSummary(aaf=AAF, allele=allele, freq=freq, coverage=coverage, loci="coding"
         , path=output_path, type="both", study="ARIC")
```

Part of the output of annotated alleles
```{r chunk 12, echo=F}
annotate_df <- read.csv(paste0(output_path, "/ARIC_annotation_variation.csv"))
annotate_df[annotate_df$mtID == 3308, ]
```

Summary of the mean coverage of loci
```{r chunk 13}
mtSum$coverLoci 
```

Summary of the mean coverage of subjects
```{r chunk 14}
mtSum$coverSubjects  
```

Display loci of variation
```{r chunk 15, eval=F}
mtSum$loci_var  
```

Summary of the heteroplasmic burden of subjects
```{r chunk 16}
mtSum$heter_burden_sum  
```

Summary of numbers of heteroplasmic mutations loci carried
```{r chunk 17}
mtSum$heter_loci_sum  
```

Display loci of heteroplasmy
```{r chunk 18, eval=F}
mtSum$loci_heter  
```

Total Number of heteroplasmic mutations
```{r chunk 19}
mtSum$heter_total  
```

Summary of the homoplasmic burden of subjects
```{r chunk 20}
mtSum$homo_burden_sum  
```

Summary of numbers of homoplasmic mutations loci carried
```{r chunk 21}
mtSum$homo_loci_sum  
```

Display loci of homoplasmy
```{r chunk 22, eval=F}
mtSum$loci_homo  
```

Total Number of homoplasmic mutations
```{r chunk 23}
mtSum$homo_total  
```

#### Annotate alternative alleles by mtAnno function

Generate alternative alleles to be annotated for mtDNA loci. It has two columns: loci positions ("pos") and "alleles" to be annotated.
```{r chunk 24}
anno <- as.data.frame(matrix(0, 10, 2))
colnames(anno)<- c("pos", "alleles")
anno$pos <- c(3311:3320)
anno$alleles <- c("A", "A", "T", "C", "A", "T", "G", "A", "C", "T")
```

Run the mtAnno function
```{r chunk 25}
mtAnno(anno=anno, path=output_path)
```

Part of the output of annotated alleles
```{r chunk 26, echo=F}
annotate_df2 <- read.csv(paste0(output_path, "/Study_mtAnno.csv"))
annotate_df2[annotate_df2$Pos == 3311, ]
```
