---
title: "ANNOmtDNA"
author: "Xianbang Sun (maintainer, sxb3000@bu.edu), Katia Bulekova, Chunyu Liu, Jessica L. Fetterman"
date: "`r format(Sys.Date(), '%m/%d/%Y')`"
abstract: >
  ANNOmtDNA is an R/BioConductor package for identifying and annotating mtDNA sequence variations. This package allows to identify mtDNA sequence variations with user-specified thresholds, and provide several plots to visualize mtDNA sequence variations. It also annotates all mtDNA variations and predicts functionality for mtDNA variants in the coding and tRNA regions. 
  ANNOmtDNA package version: `#later`

output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{ANNOmtDNA}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Standard workflow

**Note:** if you use ANNOmtDNA in published research, please cite:

> Xianbang Sun, Katia Oleinik, Chunyu Liu, Jessica L. Fetterman (2020)
> mtdnaANNO: an R package for comprehensive annotation of mtDNA sequence variation
> *Journal name*, **volume**:page.
> [link_text](http://dx.doi.org/link.address)


## Install and load the ANNOmtDNA package

```{r chunk 1}
#library(devtools)
#install_github("/ANNOmtDNA")
#library(ANNOmtDNA)
#temporary code
library(devtools)
load_all()
```

## Data preparation
Import the original allele, frequency and coverage datasets
```{r chunk 2}
header      <-  scan(paste0("/restricted/projectnb/mtdna-alcohol/Sun_Xianbang/ARIC/allele/allele.csv"),sep=",",character(),nline=1)
coverage <- matrix(scan(paste0("/restricted/projectnb/mtdna-alcohol/Sun_Xianbang/ARIC/coverage/coverage.csv"),sep=",",character()),ncol=length(header),byrow=T)
allele    <- matrix(scan(paste0("/restricted/projectnb/mtdna-alcohol/Sun_Xianbang/ARIC/allele/allele.csv"),sep=",",character()),ncol=length(header),byrow=T)
freq      <- matrix(scan(paste0("/restricted/projectnb/mtdna-alcohol/Sun_Xianbang/ARIC/freq/freq.csv"),sep=",",character()),ncol=length(header),byrow=T)
```

Assign the rownames by the column of IDs and remove the ID column
```{r chunk 3}
rownames(allele) <- allele[,1]
rownames(freq)<-freq[,1]
rownames(coverage)<-coverage[,1]
allele <- allele[,-1]
freq   <- freq[,-1]
coverage<- coverage[,-1]
```

Transpose the allele,  freq, and coverage datasets so that rows corresponds to loci and columns corresponds to subjects 
```{r chunk 4}
allele <- t(allele)
freq <- t(freq)
coverage <- t(coverage)
class(coverage) <- "numeric"
```

## Compute alternative allele fraction (AAF) by mtAAF function
```{r chunk 5}
AAF <- mtAAF(allele, freq)
```

## Scatter plot of output of mtAAF function, each point is AAF of a subject at a locus, the x axis is the mtDNA loci and the y axis is the range of AAF (0-100%)
```{r chunk 6, fig.height = 4, fig.width = 8}
plot(AAF)
```

## Plot the mean coverage of each locus by plotCover function, the x axis is the mtDNA loci and y axis is the mean coverage
```{r chunk 7, fig.height = 4, fig.width = 8}
plotCover(coverage)
```

## Histogram of the mean coverage of each subject by histSampCov function
```{r chunk 8, fig.height = 5, fig.width = 5}
histSampCov(coverage)
```

## Summarize and annotate mtDNA mutations by mtSummary function
Specify the path to output the annotation file and histograms of mtDNA mutation burden of subjects and number of mutations carried by each loci for heteroplasmic and homoplasmic mutations
```{r chunk 9}
path<-"/rprojectnb2/mtdna-alcohol/Sun_Xianbang/Annotation/Output/"
```

Run the mtSummary function, only include loci of coding region, annotate for both of heteroplasmic and homoplasmic mutations. 
```{r chunk 10}
mtSum <- mtSummary(aaf=AAF, allele=allele, freq=freq, coverage=coverage, loci="coding"
         ,path=path, type="both", study="ARIC")
```

Users can specify lower bound and upper bound of the threshold by thre.lower and thre.upper arguments. By default, thre.lower=0.03 and thre.upper=0.97. That is, if $0.03\leq AAF\leq 0.97$, it is a heteroplasmic mutation; and if $AAF>0.97$, it is a homoplasmic mutation. Users can also choose different gene regions by loci argument. For example, to choose tRNA region, set loci="tRNA". Users can also choose types of mutations to be annotated by type argument. For example, to choose heteroplasmic mutations, set type="heter", and choose homoplasmic mutations, set type="homo".

Summary of the mean coverage of loci
```{r chunk 11}
mtSum$coverLoci 
```

Summary of the mean coverage of subjects
```{r chunk 12}
mtSum$coverSubjects  
```

Display loci of variation
```{r chunk 13, eval=F}
mtSum$loci_var  
```

Summary of the heteroplasmic burden of subjects
```{r chunk 14}
mtSum$heter_burden_sum  
```

Summary of numbers of heteroplasmic mutations loci carried
```{r chunk 15}
mtSum$heter_loci_sum  
```

Display loci of heteroplasmy
```{r chunk 16, eval=F}
mtSum$loci_heter  
```

Total Number of heteroplasmic mutations
```{r chunk 17}
mtSum$heter_total  
```

Summary of the homoplasmic burden of subjects
```{r chunk 18}
mtSum$homo_burden_sum  
```

Summary of numbers of homoplasmic mutations loci carried
```{r chunk 19}
mtSum$homo_loci_sum  
```

Display loci of homoplasmy
```{r chunk 20, eval=F}
mtSum$loci_homo  
```

Total Number of homoplasmic mutations
```{r chunk 21}
mtSum$homo_total  
```

## Annotate alternative alleles by mtAnno function

Generate alternative alleles to be annotated for mtDNA loci. It has two columns: loci positions ("pos") and "alleles" to be annotated.
```{r chunk 22}
anno <- as.data.frame(matrix(0, 10, 2))
colnames(anno)<- c("pos","alleles")
anno$pos <- c(3311:3320)
anno$alleles <- c("A", "A", "T", "C", "A", "T", "G", "A", "C", "T")
```

Run the mtAnno function
```{r chunk 23}
mtAnno(anno=anno, path=path)
```






